### PR TITLE
Compatibility with Pytest 5.4

### PR DESCRIPTION
**Pull request by saimn**
This is quite similar to #95, but I started from scratch and porting some code from pytest's doctest plugin which probably changed a lot since the first version of doctestplus.

With these changes I had a few tests failing, and I actually don't understand why the tests were not failing before so I updated the tests. Please review, not sure who knows this code best ?

### CODE REVIEW

File: tests/test_doctestplus.py [hunk 1]
```diff
@@ -351,7 +351,7 @@ def test_requires(testdir):
             >>> import foobar
         """
     )
-    testdir.inline_run(p, '--doctest-plus', '--doctest-rst').assertoutcome(passed=1)
+    testdir.inline_run(p, '--doctest-plus', '--doctest-rst').assertoutcome(skipped=1)
```
Comment (saimn) [inline tests/test_doctestplus.py:354]: This is testing that the test is skipped because of a missing import, I guess the outcome should be "skipped".

File: tests/test_doctestplus.py [hunk 2]
```diff
@@ -394,7 +394,7 @@ def test_requires(testdir):
         """
     )
     # passed because 'pytest<1.0' was not satisfied and 'assert 0' was not evaluated
-    testdir.inline_run(p, '--doctest-plus', '--doctest-rst').assertoutcome(passed=1)
+    testdir.inline_run(p, '--doctest-plus', '--doctest-rst').assertoutcome(skipped=1)
```
Comment (saimn) [inline tests/test_doctestplus.py:397]: Same here.

File: tests/test_doctestplus.py [hunk 3]
```diff
@@ -510,7 +510,7 @@ def test_text_file_comments(testdir):
         '--doctest-glob', '*.rst',
         '--doctest-glob', '*.tex',
         '--doctest-glob', '*.txt'
-    ).assertoutcome(passed=3)
+    ).assertoutcome(passed=0)
```
Comment (saimn) [inline tests/test_doctestplus.py:513]: This is testing that the tests are commented, so they are not run, they should not "pass". Same below.

File: tests/test_doctestplus.py (hunk 1 above)
Comment (bsipocz) [reply tests/test_doctestplus.py:354]: how did all of these tests work before?

File: pytest_doctestplus/plugin.py [hunk 4]
```diff
@@ -205,40 +204,39 @@ def collect(self):
...
+            text = self.fspath.read_text(encoding)
+            filename = str(self.fspath)
+            name = self.fspath.basename
+            globs = {"__name__": "__main__"}
 
-            doctest.testfile(
-                str(self.fspath), module_relative=False,
-                optionflags=options, parser=DocTestParserPlus(),
-                extraglobs=dict(getfixture=fixture_request.getfixturevalue),
-                raise_on_error=True, verbose=False, encoding='utf-8')
```
Comment (saimn) [inline pytest_doctestplus/plugin.py:229]: Ok so the tests differences are due to this removal of `doctest.testfile`, which is because I basically copied what pytest's doctest plugin is doing, to remove the multiple inheritance, and use `collect` instead of `runtest`. I don't why the old code was doing it another way so ... :man_shrugging:

File: tests/test_doctestplus.py (hunk 1 above)
Comment (saimn) [reply tests/test_doctestplus.py:354]: See my comment below: https://github.com/astropy/pytest-doctestplus/pull/103#pullrequestreview-385227537

**Comment by bsipocz** [comment]
There are still issues with pytest><4 versions. Those are however 1.5 years old, so we may even consider dropping them if the fix is more complicated than necessary.

**Comment by saimn** [comment]
Oh sorry, I thought I read somewhere that pytest's minversion was 5.0, but I must have mistaken it for another package. Anyway, for tests dependencies I think it is easier to require recent versions, so indeed I'd be happy if we can drop old versions.

**Comment by bsipocz** [comment]
ok, let's wait for @astrofrog's opinion. And nominally @eteq is also listed as testing maintainer, in case he wants to chime in...

**Comment by bsipocz** [comment]
@saimn - Shall I open a separate PR for the travis updates, or push directly to your branch? There are other unrelated issues with the config, so would ideally all need to be cleaned up before this is merged (to be sure this works).

**Comment by saimn** [comment]
@astrofrog - yes it works fine with astropy. The test cases I had to changes were edge cases, and I mostly copied code from pytest's doctest plugin, so hopefully it should be fine.

@bsipocz - sure, push here if you want. The travis config indeed needs some updates.

**Comment by bsipocz** [comment]
This now drops support for python <3.6 as well as for pytest <4.

**Comment by bsipocz** [comment]
I've also added a job for pytest dev version, and switched on the weekly cron.

Now this is very much ready for merging.

**Comment by bsipocz** [comment]
No vetoes here in the past week, so I'm merging and will go ahead with a release fairly soon.

Thank you @saimn  for picking this up and fixing it.

File: pytest_doctestplus/plugin.py [hunk 5]
```diff
@@ -164,24 +163,23 @@ def collect(self):
...
                 try:
                     module = self.fspath.pyimport()
-                    # Just ignore searching modules that can't be imported when
-                    # collecting doctests
                 except ImportError:
-                    return
+                    if self.config.getvalue("doctest_ignore_import_errors"):
+                        pytest.skip("unable to import module %r" % self.fspath)
+                    else:
+                        raise
```
Comment (jdavies-st) [inline pytest_doctestplus/plugin.py:174]: This changes dramatically the behavior of the plugin.  Now if any module in the repo using this plugin has import failures, it will cause a collection error.  In the past for this plugin, this was ignored.  I see that using `--doctest-ignore-import-errors` can turn this off, but it would be good to document this in the README and the release notes.

Comment (astrofrog) [reply pytest_doctestplus/plugin.py:174]: See https://github.com/astropy/astropy/pull/10242 for a possible fix in astropy btw, we should discuss there whether that is the correct fix.

Comment (saimn) [reply pytest_doctestplus/plugin.py:174]: Hmm, sorry for the breakage. The goal was just to reduce the number of difference with pytest's doctest plugin, because of lot of the code here is copied from there, and reducing the differences should make it easier to maintain... 
But yeah as previously this was ignored by default it is probably easier to fix this here.

Comment (jdavies-st) [reply pytest_doctestplus/plugin.py:174]: No worries.  I think it is easy to add `--doctest-ignore-import-errors` to addopts in setup.cfg, it was just surprising.  And it took some time to debug.  Hence the desire of making this clear in the documentation.

Comment (jdavies-st) [reply pytest_doctestplus/plugin.py:174]: Also, it was nice finding out we had tumbleweed code laying about that had imports that didn't work.

I agree that having its behavior similar to pytest's `doctest` is good.  The more things that are the same, not only easier to maintain, but the less unexpected behavior for people using the plugin as an extension or replacement of the pytest one.

Comment (bsipocz) [reply pytest_doctestplus/plugin.py:174]: see a few open issues about moving these upstream and consolidate the differences between doctest and doctestplus. The blocking factor is time, and now I also think it's better to wait for the situation at pytest to resolve.

Comment (bsipocz) [reply pytest_doctestplus/plugin.py:174]: also, it's my bad that I didn't pick this up during the review, I suppose I put too much trust in tests and was just too happy when this PR finally had a green status.

**Comment by dstansby** [comment]
Has this made its way into a release yet?

**Comment by saimn** [comment]
@dstansby - Yes, in 0.6 : https://github.com/astropy/pytest-doctestplus/blob/master/CHANGES.rst

**Comment by dstansby** [comment]
Thanks!
